### PR TITLE
Samples generating code for Convolutional VAE example

### PR DESCRIPTION
**Pull request by hello-fri-end**
The Convolutional VAE example in the models repo had some extra code for testing & generating samples which I've copied in this PR. I'll open a PR in the models repo with the segfault fix later today.

### CODE REVIEW

File: mnist_vae_cnn/mnist_vae_cnn.cpp [hunk 1]
```diff
@@ -84,6 +86,13 @@ int main()
 
   arma::mat train, validation;
   data::Split(fullData, validation, train, trainRatio);
+  
+  // Loss is calculated on train_test data after each cycle.
+  arma::mat train_test, dump;
```
Comment (zoq) [inline mnist_vae_cnn/mnist_vae_cnn.cpp:91]: Let's use camel casing for all names.

File: mnist_vae_cnn/mnist_vae_cnn.cpp [hunk 2]
```diff
@@ -84,6 +86,13 @@ int main()
...
+  data::Split(train, dump, train_test, 0.045);
+
+  // No of iterations of the optimizer
```
Comment (zoq) [inline mnist_vae_cnn/mnist_vae_cnn.cpp:94]:
```suggestion
  // No of iterations of the optimizer.
```

File: mnist_vae_cnn/mnist_vae_cnn.cpp [hunk 3]
```diff
@@ -197,24 +206,42 @@ int main()
...
       0.999, // Exponential decay rate for the weighted infinity norm estimates.
       1e-8,  // Value used to initialise the mean squared gradient parameter.
-      maxIteration, // Max number of iterations.
+      iterPerCycle, // Max number of iterations.
       1e-8,         // Tolerance.
       true);
+      
+  const clock_t begin_time = clock();
+  clock_t cycle_time = begin_time;
+
```
Comment (zoq) [inline mnist_vae_cnn/mnist_vae_cnn.cpp:215]:
```suggestion
```

Just removed the extra line here.

File: mnist_vae_cnn/mnist_vae_cnn.cpp [hunk 4]
```diff
@@ -197,24 +206,42 @@ int main()
...
+     		   ens::ProgressBar());
+     		   
+    // Don't reset optimizer's parameters between cycles.
+    optimizer.ResetPolicy() = false;
+
+    std::cout << "Loss after cycle  " << i << " -> " <<
+        MeanTestLoss<MeanSModel>(vaeModel, train_test, batchSize) << std::endl;
+    std::cout << "Time taken for cycle -> " << float(clock() - cycle_time) /
+        CLOCKS_PER_SEC << " seconds" << std::endl;
+    cycle_time = clock();
```
Comment (zoq) [inline mnist_vae_cnn/mnist_vae_cnn.cpp:235]: We could also use the `Report` callback - https://ensmallen.org/docs.html#report for this one, not sure the output is too much, let me know what you think.

File: mnist_vae_cnn/vae_generate.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  constexpr bool isBinary = false;
+  // the latent size of the VAE model.
+  constexpr size_t latentSize = 20;
+
+  arma::mat fullData, train, validation;
+
+  if (loadData)
+  {
+    data::Load("../data/mnist_train.csv", fullData, true, false);
+    // Get rid of the header
```
Comment (zoq) [inline mnist_vae_cnn/vae_generate.cpp:50]:
```suggestion
    // Get rid of the header.
```

File: mnist_vae_cnn/vae_generate.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    fullData = 
+        fullData.submat(0, 1, fullData.n_rows - 1, fullData.n_cols -1);
+    fullData /= 255.0;
+    // Get rid of the labels
```
Comment (zoq) [inline mnist_vae_cnn/vae_generate.cpp:54]:
```suggestion
    // Get rid of the labels.
```

File: mnist_vae_cnn/mnist_vae_cnn.cpp (hunk 4 above)
Comment (hello-fri-end) [reply mnist_vae_cnn/mnist_vae_cnn.cpp]: Strangely, on adding the `Report `callback, I get an error saying " `Report` is not a part of 'ens' ". Maybe, the version of ensmallen needs to be updated in the `makefile`, not sure. .

Comment (zoq) [reply mnist_vae_cnn/mnist_vae_cnn.cpp]: What version of ensmallen do you us; `Report` was added in https://github.com/mlpack/ensmallen/blob/master/HISTORY.md#ensmallen-2140-no-direction-home

Comment (hello-fri-end) [reply mnist_vae_cnn/mnist_vae_cnn.cpp]: Yup, my ensmallen needed an update. This is a pretty cool callback. I removed `cycleTime` as this callback already calculates that for us.

**Review by zoq — APPROVED** [review]
Looks great to me, thanks for working on this one.

**Pull request merged**